### PR TITLE
Added LivingVisibilityEvent as a replacement for PlayerEvent.Visibility

### PR DESCRIPTION
**Pull request by apace100**
In 1.12.x, we had the PlayerEvent.Visibility to allow mods to change the modifier that is applied to the distance limit the AI uses when it determines whether a mob can see the player.

That event has not been hooked up yet in 1.14.x.

This PR adds a new, more general, LivingVisibilityEvent, which is called not only when the player's visibility is checked, but also for other living entities (such as sheep which might be seen by wolves).

Additionally, the new event includes the (nullable) entity which is performing the visibility check, which allows modders to react to it.

The old PlayerEvent.Visibility is now called again for players, but is marked as deprecated and suggests moving to the new event.

The PR includes a test mod which adds the following functionality:
- Players which wear a leather chestplate are not seen by skeletons
- Players which wear a golden chestplate are seen from all mobs from 4 times as far away
- A debug call which shows that the old PlayerEvent.Visibility event is called

### CODE REVIEW

File: src/main/java/net/minecraftforge/common/ForgeHooks.java [hunk 1]
```diff
@@ -374,7 +375,18 @@ public static int getLootingLevel(LivingEntity target, DamageSource cause, int l
...
     }
-
+    
+    public static double getLivingVisibility(LivingEntity living, @Nullable Entity lookingEntity, double visibility)
+    {
+    	LivingVisibilityEvent event = new LivingVisibilityEvent(living, lookingEntity, visibility);
+        MinecraftForge.EVENT_BUS.post(event);
+        double value = event.getVisibility();
+        if(living instanceof PlayerEntity) {
+            value = getPlayerVisibilityDistance((PlayerEntity)living, value, visibility);
```
Comment (ichttt) [inline src/main/java/net/minecraftforge/common/ForgeHooks.java]: I think the legacy event should be fired first, so the new event can overwrite the legacy result

File: src/main/java/net/minecraftforge/common/ForgeHooks.java [hunk 2]
```diff
@@ -374,7 +375,18 @@ public static int getLootingLevel(LivingEntity target, DamageSource cause, int l
         MinecraftForge.EVENT_BUS.post(event);
         return event.getLootingLevel();
     }
-
+    
+    public static double getLivingVisibility(LivingEntity living, @Nullable Entity lookingEntity, double visibility)
+    {
+    	LivingVisibilityEvent event = new LivingVisibilityEvent(living, lookingEntity, visibility);
```
Comment (ichttt) [inline src/main/java/net/minecraftforge/common/ForgeHooks.java]: Wrong indent

**Review by ichttt — CHANGES_REQUESTED** [review]
Two small nitpicks, otherwise fine

**Comment by LexManos** [comment]
Honestly, this doesn't look like something that should be an event, but perhaps introduce a new DataAttribute? That way the value can be computed once when state changes {crouching, armor change, daytime change, etc..} instead of every time it's called?

As for the old event, that needs a `@Deprecated//Never used, Remove 1.15 added` so we remember to remove it.

**Comment by apace100** [comment]
> Honestly, this doesn't look like something that should be an event, but perhaps introduce a new DataAttribute? That way the value can be computed once when state changes {crouching, armor change, daytime change, etc..} instead of every time it's called?

I'm not sure whether an attribute is the right choice here - sure, it eases the computation a bit, and I agree that the event in its current form is quite expensive because it gets fired many times per frame (potentially). But moving it to an attribute would remove the flexibility for modders to react to the watching entity, e.g. to reduce visibility only for skeletons. I think we should keep that possibility, because Vanilla does similar things (wearing a zombie skull and a zombie is watching? visibility *= 0.5).
If we don't deem the flexibility valuable enough we could of course move the base value to an attribute, and then simply let vanilla apply modifiers based on the watching entity.
What do you think?

> As for the old event, that needs a `@Deprecated//Never used, Remove 1.15 added` so we remember to remove it.

Sure, I can do that.

**Pull request closed without merging**